File: NTUA-Programming-Languages-I/Lab 2/fairmaze/fairmaze Python/fairmaze.py
import sys
import numpy as np
from collections import deque

MAXN = 1005
MAXM = 1005

adj = {}
visited = set()
component = {}

def is_out(x, y):
    if x>=0 and  x <= N-1 and y >= 0 and y <= M-1 :
        return False
    else:
        return True

def add_edge(x1, y1, x2, y2):
    if is_out(x2, y2):
        x2 = N
        y2 = M

    # we save the edge reversed
    if (x2,y2) not in adj:
        adj[(x2,y2)] = []

    adj[(x2,y2)].append((x1,y1))

def bfs(x, y):
    visited.add((x,y))
    Q = deque([(x,y)])

    while(len(Q)!=0):
        v = Q.popleft()

        z = 0 if v not in adj else len(adj[v])

        for i in range(z):
            x = adj[v][i]

            if x not in visited:
                visited.add(x)
                Q.append(x)


file = sys.argv[1]
f = open(file,"r")
s = f.read()
lines = s.split('\n')[:-1]

N, M = [int(x) for x in lines[0].split(' ')]

for i in range(N):
    for j in range(M):

        c = lines[i+1][j]

        if c == 'L':
            add_edge(i,j,i,j-1)
        elif c == 'R':
            add_edge(i,j,i,j+1)
        elif c == 'D':
            add_edge(i,j,i+1,j)
        elif c == 'U':
            add_edge(i,j,i-1,j)

visited.add((N,M))

bfs(N, M)

result = 0

for i in range(N):
    for j in range(M):
        if (i,j) not in visited:
            result += 1

print(result)

File: NTUA-Programming-Languages-I/Lab 1/fairmaze/fairmaze C++/fairmaze.cpp
#include <iostream>
#include <fstream>
#include <vector>

#define MAXN 1005
#define MAXM 1005

using namespace std;

int N, M;
char c;
vector<pair<int,int>> adj[MAXN][MAXM];
int visited[MAXN][MAXM];

bool is_out(int x, int y){
  if( x >= 0 && x <= N-1  && y >= 0 && y <= M-1 ) return false;
  return true;
}

void add_edge(int x1, int y1, int x2, int y2){

  if(is_out(x2,y2)){
    x2 = N;
    y2 = M;
  }

  // we save the edge reversed
  adj[x2][y2].push_back({x1,y1});
}

void dfs(int x, int y){
  visited[x][y] = 1;


  for(int i = 0; i < (int) adj[x][y].size(); i++){

    int xNext = adj[x][y][i].first;
    int yNext = adj[x][y][i].second;

    if( !visited[xNext][yNext] ){
      dfs(xNext, yNext);
    }
  }
}

int main(int argc, char* argv[]){
  ifstream inputFile;
  inputFile.open(argv[1]);

  inputFile >> N >> M;

  for(int i = 0; i < N; i++){
    for(int j = 0; j < M; j++){
      inputFile >> c;
      switch (c) {
        case 'L':
          add_edge(i,j,i,j-1);
          break;
        case 'R':
          add_edge(i,j,i,j+1);
          break;
        case 'D':
          add_edge(i,j,i+1,j);
          break;
        case 'U':
          add_edge(i,j,i-1,j);
          break;
      }
    }
  }

  visited[N][M] = 1;

  dfs(N,M);

  int result = 0;

  for(int i = 0; i < N; i++){
    for(int j = 0; j < M; j++){
      if(visited[i][j] == 0){
        result++;
      }
    }
  }

  cout << result << endl;

}